Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,224 @@
+            if let data = controlMessage.data {
+                assert(data.count == 1)
+                precondition(data.count >= 1)
+                let readValue: Int32 = .init(data[0])
+                self.ecnValue = ControlMessageReceiver.parseEcn(receivedValue: readValue)
+            }
+        } else if controlMessage.level == IPPROTO_IPV6 && controlMessage.type == IPV6_TCLASS {
+            if let data = controlMessage.data {
+                assert(data.count == 4)
+                precondition(data.count >= 4)
+                var readValue: Int32 = 0
+                withUnsafeMutableBytes(of: &readValue) { valuePtr in
+                    valuePtr.copyMemory(from: data)
+                }
+                self.ecnValue = ControlMessageReceiver.parseEcn(receivedValue: readValue)
+            }
+        }
+    }
+
+    private static func parseEcn(receivedValue: Int32) -> NIOExplicitCongestionNotificationState {
+        switch receivedValue & IPTOS_ECN_MASK {
+        case IPTOS_ECN_ECT1:
+            return .transportCapableFlag1
+        case IPTOS_ECN_ECT0:
+            return .transportCapableFlag0
+        case IPTOS_ECN_CE:
+            return .congestionExperienced
+        default:
+            return .transportNotCapable
+        }
+    }
+}
+
+extension NIOExplicitCongestionNotificationState {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    fileprivate static let notCapableValue = IPTOS_ECN_NOTECT
+    #else
+    fileprivate static let notCapableValue = IPTOS_ECN_NOT_ECT    // Linux
+    #endif
+    
+    func asCInt() -> CInt {
+        switch self {
+        case .transportNotCapable:
+            return .init(NIOExplicitCongestionNotificationState.notCapableValue)
+        case .transportCapableFlag0:
+            return .init(IPTOS_ECN_ECT0)
+        case .transportCapableFlag1:
+            return .init(IPTOS_ECN_ECT1)
+        case .congestionExperienced:
+            return .init(IPTOS_ECN_CE)
+        }
+    }
+}
+
+struct UnsafeOutboundControlBytes {
+    private var controlBytes: UnsafeMutableRawBufferPointer
+    private var writePosition: size_t = 0
+    
+    /// Control bytes are captured - this structure must not have a lifetime exceeded them.
+    init(controlBytes: UnsafeMutableRawBufferPointer) {
+        self.controlBytes = controlBytes
+    }
+    
+    mutating func appendControlMessage<PayloadType>(level: Int32,
+                                          type: Int32,
+                                          payload: PayloadType) {
+        let writableBuffer = UnsafeMutableRawBufferPointer(rebasing: self.controlBytes[writePosition...])
+        
+        let requiredSize = Posix.cmsgSpace(payloadSize: MemoryLayout.stride(ofValue: payload))
+        precondition(writableBuffer.count >= requiredSize, "Insufficient size for cmsghdr and data")
+        
+        let bufferBase = writableBuffer.baseAddress!
+        let cmsghdrPtr = bufferBase.bindMemory(to: cmsghdr.self, capacity: 1)
```

Comment:
Any use of the memory binding APIs requires considerable caution and some commenting to explain why it's fine (or indeed, not fine).

In this case, we want the bound pointer because we need it for `cmsgData`, so we need to bind it. However, we drop the bound pointer after this function call. As we own the original pointer, if we're confident we never bind this to anything _else_ at the same time we're good to go. We know we don't (no other bound pointer to this memory can exist at the same time), so we're safe, but we need to write all of this down.